multiuser/api/che-multiuser-api-authentication-commons/src/main/java/org/eclipse/che/multiuser/api/authentication/commons/filter/SessionCachingFilter.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.multiuser.api.authentication.commons.filter;
+
+import javax.servlet.Filter;
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import javax.servlet.http.HttpSession;
+import org.eclipse.che.multiuser.api.authentication.commons.SessionStore;
+
+public abstract class SessionCachingFilter implements Filter {
+
+  private SessionStore sessionStore;
+
+  public SessionCachingFilter(SessionStore sessionStore) {
+    this.sessionStore = sessionStore;
+  }
+
+  protected class SessionCachedHttpRequest extends HttpServletRequestWrapper {
+
+    /**
+     * Constructs a request object wrapping the given request.
+     *
+     * @throws IllegalArgumentException if the request is null
+     */
+    public SessionCachedHttpRequest(ServletRequest request) {
+      super((HttpServletRequest) request);
+    }
+
+    @Override
+    public HttpSession getSession() {
+      return getOrCreateSession(true);
+    }
+
+    @Override
+    public HttpSession getSession(boolean create) {
+      return getOrCreateSession(create);
+    }
+
+    private HttpSession getOrCreateSession(boolean createNew) {
```

Comment:
could you please javadoc this method ? It has quite complicated logic. I would consider to split it to multiple methods.